keras_nlp/samplers/__init__.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from tensorflow import keras
+
+from keras_nlp.samplers.greedy import Greedy
+
+
+def serialize(sampler):
+    return keras.utils.serialize_keras_object(sampler)
+
+
+def deserialize(config, custom_objects=None):
+    """Return a `Sampler` object from its config."""
+    all_classes = {
+        "greedy": Greedy,
+    }
+    if config["class_name"].lower() in all_classes:
+        config["class_name"] = config["class_name"].lower()
+    return keras.utils.deserialize_keras_object(
+        config,
+        module_objects=all_classes,
+        custom_objects=custom_objects,
+        printable_module_name="samplers",
+    )
+
+
+def get(identifier):
+    """Retrieve a KerasNLP sampler by the identifier.
+
+    The `identifier` may be the string name of a sampler class or class.
+
+    >>> identifier = 'greedy'
+    >>> sampler = keras_nlp.samplers.get(identifier)
+
+    You can also specify `config` of the sampler to this function by passing
+    dict containing `class_name` and `config` as an identifier. Also note that
+    the `class_name` must map to a `Sampler` class.
+
+    >>> cfg = {'class_name': 'Greedy', 'config': {}}
+    >>> sampler = keras_nlp.samplers.get(cfg)
+
+    In the case that the `identifier` is a class, this method will return a new
+    instance of the class by its constructor.
+
+    Args:
+      identifier: String or dict that contains the sampler name or
+        configurations.
+
+    Returns:
+      Sampler instance base on the input identifier.
+
+    Raises:
+      ValueError: If the input identifier is not a supported type or in a bad
+        format.
+    """
+
+    if identifier is None:
+        return None
+    if isinstance(identifier, dict):
+        return deserialize(identifier)
+    elif isinstance(identifier, str):
+        identifier = {"class_name": str(identifier), "config": {}}
```

Comment:
this line can actually just be deleted!

```
keras.utils.deserialize_keras_object("greedy", module_objects={
    "greedy": Greedy,
})
```

just works

Reply:
I went a step further to allow mixed case, if users pass "Greedy", we still let it work. This is the same pattern as optimizer, e.g., "Adam"/"ADAM"/"adam" gives the same thing, which I find useful for users. I force the `deserialize()` to take in a dict, so the lowercase code is unifed
```
if config["class_name"].lower() in all_classes:
    config["class_name"] = config["class_name"].lower()
```

Reply:
did not see the comment below, ignore this

Reply:
Yeah, I just think that allowing more forms is a bad thing. It is better if the KerasNLP code out there is uniform `sampler="greedy"`, why not force that?

For optimizers and initializers we have to support these variants for backwards compat. But here we can be strict!